### ISSUE TITLE
Occasional errors connecting to the cloudsql proxy

### ISSUE THREAD

**Issue by josephtyler** [GoogleCloudPlatform/cloudsql-proxy#184]
In my cloudsql logs I frequently see errors like this:

```
E  2018/06/27 06:18:50 New connection for "myproject:region:db"
E  2018/06/27 06:18:51 Client closed local connection on 127.0.0.1:5432
```

And in my application logs I see this:

```
OperationalError at /url
could not connect to server: Connection refused
        Is the server running on host "127.0.0.1" and accepting
        TCP/IP connections on port 5432?
```

I'm not really sure how to go about troubleshooting this. This error doesn't happen consistently, seems to happen in bursts of 10-15 a few times a day.

**Comment by kevin-jj** [comment]
I was running into same issue on app server boot and have no idea to solve it. @Carrotman42

**Comment by kurtisvg** [comment]
@josephtyler Is this still an issue for you? Can you provide any more information or an example of the code you are using to connect?

**Comment by rahmonov** [comment]
I am having the same issue since we migrated to CloudSQL. Here is more info:

![connection refused](https://user-images.githubusercontent.com/9026972/43064556-ae478674-8e78-11e8-92d7-26277eef8cb6.png)

**Comment by hfwang** [comment]
If it's happening sporadically and the error is connection refuse, I'd expect to see the cloud sql proxy getting restarted or something else happening to it...

@josephtyler 
What environment are you running in?

When you say cloudsql logs, do you mean the logs for the cloud sql proxy or for the instance itself? Is there nothing else in the logs?

Also, just to double check: you are running the cloud sql proxy and binding it to 127.0.0.1:5432 right?

@rahmonov 
Are you able to share any logs from the cloud sql proxy?

**Comment by fernandonogueira** [comment]
This is also happening here.
The proxy says it is ready to accept new connections but my application is throwing a timeout exception.

I followed the instructions from this page to connect my java/spring application to an Cloud SQL instance.

https://cloud.google.com/sql/docs/postgres/connect-kubernetes-engine

**Comment by fernandonogueira** [comment]
Hello guys!

In fact my problem was quite different.
I needed to add a sleep before starting my application that was running in the same pod. 

After that, the proxy started receiving connections from my Spring application.

Thanks!

**Comment by kurtisvg** [comment]
Going to go ahead and close this - it's been over 30 days since it was first opened and there isn't enough information to verify any issues.

**Comment by rahmonov** [comment]
@kurtisvg @hfwang I think I found out why I am having this errors:

When a pod is terminated during rollout, the CloudSQL proxy container is terminated first and only after some seconds my application container is terminated. So, before my container is terminated, it keeps sending requests to the CloudSQL container which is already down, resulting in such errors.

If I could somehow specify the order of termination for the containers inside a single pod, this would solve this issue, I think.

I am trying to do something with [container life cycle events](https://kubernetes.io/docs/tasks/configure-pod-container/attach-handler-lifecycle-event/), will write here if I solve it.

**Comment by kurtisvg** [comment]
@rahmonov You may be interested in #178 as well

**Comment by rahmonov** [comment]
@kurtisvg Thank you! Before it is merged and released, I am solving this by adding a `preStop` lifecycle event:

```
lifecycle:
  preStop:
    exec:
      command: ["/bin/sleep", "300"]
```

I am just having my cloudsql proxy container sleep until my application is down (which is 30s by default)

**Comment by markvincze** [comment]
Hi @josephtyler,

Did you manage to find the root cause and a solution for the original issue?  
I'm seeing the exact same problem (`New connection for...` message in the Cloud SQL Proxy logs, and a `Connection refused` error in my aplpication), and it's definitely not happening either during the startup nor the termination of a pod, but during normal operation.

**Comment by javapapo** [comment]
+1 on the above version 1.15

**Comment by pslobo** [comment]
I'm seeing the exact same situation as @markvincze. In a 30s load test, I'll see loads of `New connection for...` despite using connection pooling in the code. This, plus the added overhead of establishing new connections that Cloud Proxy introduces means that I can't even max out the low cpu/memory limits I have on my pods before starting to get status 500 from my services due to DB connection issues.

**Comment by elmarti** [comment]
@pslobo I'm finding the exact same issue - excluding Istio traffic helped somewhat with this, but even with 40rps, i'm finding that my DB connection goes away. Ample mem and CPU for both the app and the database.

**Comment by pslobo** [comment]
@elmarti we eventually had to give up on using the proxy and establish direct connections using the private IP. Performance has increased significantly at the expense of not encrypting traffic via proxy

**Comment by enocom** [comment]
We've made a number of improvements since v1.17. I'd be very curious to hear if anyone is experiencing this problem in the latest version [v1.28.0](https://github.com/GoogleCloudPlatform/cloudsql-proxy/releases/tag/v1.28.0).

**Comment by bugraoz93** [comment]
Hi @enocom, we have updated our version to [v1.28.0](https://github.com/GoogleCloudPlatform/cloudsql-proxy/releases/tag/v1.28.0). Unfortunately, we are still facing the same error that was mentioned by @markvincze. We are seeing this error (`New connection for...`) in the console logs up to 2 or 3 times in a minute.

**Comment by enocom** [comment]
To be clear, logs about clients opening and closing connections are totally normal as clients come and go.

If the application is dropping connections, that's another problem that could be caused by all sorts of application-side issues.

@bugraoz93 Would you mind opening a new issue, providing details on your setup and what problems you're seeing? This issue is no longer focused on a single problem.

**Comment by bugraoz93** [comment]
Thank you for the information and the quick answer @enocom. I am very new at CloudSQL. I wanted to raise the problem since you have asked if this issue persists or not. I thought it could be related. As you mention, if the connection initiation and termination logs are info level logs then it is okay. 

In one of our environments, our system works without a problem but with the same system infrastructure in a different project (that uses GCP infrastructure), we have the same connection problem same as this which is dropping the connection occasionally (in random times) and cannot continue to communicate again from that point. This led to the same connection initiation and termination logs multiple times in a row.

You are right, the problem could be related to the application side. After ensuring it is not, I will create another issue that will include our infrastructure detail and the error detail.

**Comment by enocom** [comment]
Thanks, @bugraoz93. Even if it's just a question, we do have a "question" type for issues. Feel free to open those as well.

**Comment by Tech2pr** [comment]
@enocom We are occasionally getting the connection refused error in production. We get this error 2-3 times a day. Our app is deployed in App Engine. I couldn't find anything helpful in the logs. We are using  psycopg2 v 2.8.5 binary as postgresql engine.

could not connect to server: Connection refused
        Is the server running locally and accepting
        connections on Unix domain socket "/cloudsql/xxxxxxxx:us-central1:xxxxx-production-1/.s.PGSQL.5432"?

**Comment by enocom** [comment]
@Tech2pr Would you mind opening a new issue? This one is a bunch of unrelated items.
